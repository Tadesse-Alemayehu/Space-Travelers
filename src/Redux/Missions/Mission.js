const STORE_MISSIONS = 'SpaceTravelers/Mission/STORE_MISSIONS';
const RESERVATION = 'SpaceTravelers/Mission/RESERVATION';
const initialState = [];

export const reserveMission = (id, status) => (
  {
    type: RESERVATION,
    mission: {
      id,
      status: Boolean(Number(status)),
    }
  }
);

export const getMissions = () => async (dispatch) => {
  fetch('https://api.spacexdata.com/v3/missions')
    .then(data => data.json())
    .then(data => {
      console.log(data);
      dispatch({ type: STORE_MISSIONS, data });
    })
    .catch(error => console.log(error));
};

const missionsReducer = (state = initialState, action) => {
  switch (action.type) {
    case STORE_MISSIONS:
      return action.data;

    case RESERVATION:
      return [...state.map(mission => {
        if (mission.mission_id === action.mission.id)
          return { ...mission, reserved: action.mission.status }
        return mission;
      }
      )];

    default:
      return state;
  }
};

export default missionsReducer;
