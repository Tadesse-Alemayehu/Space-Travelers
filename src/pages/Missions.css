table {
  margin: 4vw;
  border-collapse: collapse;
  text-align: left;
}

table th {
  padding: 0.5vw;
  border: 1px solid #dee2e6;
}

table td {
  border: 1px solid #dee2e6;
}

table tbody tr:nth-child(odd) {
  background-color: #f2f2f2;
}

table tbody tr:nth-child(even) {
  background-color: white;
}

.join-mission {
  padding: 0 3vw;
}

.member-mission {
  padding: 0 0.5vw;
}

.mission-title {
  vertical-align: top;
  padding: 0.5vw;
  font-weight: bold;
}

.mission-paragraph {
  font-size: 0.6em;
  line-height: 1.5em;
  padding: 0.5vw 0.5vw 3vw;
}

.member {
  -webkit-appearance: none;
     -moz-appearance: none;
          appearance: none;
  background: none;
  border: none;
  background-color: #6d757d;
  color: white;
  border-radius: 4px;
  display: -webkit-box;
  display: -ms-flexbox;
  display: flex;
  width: -webkit-max-content;
  width: -moz-max-content;
  width: max-content;
  font-weight: bold;
  padding: 0.2vw 0.5vw;
}

.active-member {
  -webkit-appearance: none;
     -moz-appearance: none;
          appearance: none;
  background: none;
  border: none;
  background-color: #2ba8be;
  color: white;
  border-radius: 4px;
  display: -webkit-box;
  display: -ms-flexbox;
  display: flex;
  width: -webkit-max-content;
  width: -moz-max-content;
  width: max-content;
  font-weight: bold;
  padding: 0.2vw 0.5vw;
}

.join {
  -webkit-appearance: none;
     -moz-appearance: none;
          appearance: none;
  background: none;
  border: none;
  border: 1px solid #343a40;
  color: #343a40;
  border-radius: 3px;
  width: -webkit-max-content;
  width: -moz-max-content;
  width: max-content;
  padding: 0.7vw 0.5vw;
  cursor: pointer;
}

.leave {
  -webkit-appearance: none;
     -moz-appearance: none;
          appearance: none;
  background: none;
  border: none;
  border: 1px solid #dc3645;
  color: #dc3645;
  border-radius: 3px;
  width: -webkit-max-content;
  width: -moz-max-content;
  width: max-content;
  padding: 0.7vw 0.5vw;
  cursor: pointer;
}
/*# sourceMappingURL=Missions.css.map */