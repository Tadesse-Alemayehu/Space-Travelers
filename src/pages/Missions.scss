table {
  margin: 4vw;
  border-collapse: collapse;
  text-align: left;

  th {
    padding: 0.5vw;
    border: 1px solid #dee2e6
  }

  td {
    padding: 0.5vw 0.5vw 3vw;
    border: 1px solid #dee2e6;
    vertical-align: top;
  }

  tbody tr:nth-child(odd) {
    background-color: #f2f2f2;
  }
  
  tbody tr:nth-child(even) {
    background-color: white;
  }
}

.mission-title {
  vertical-align: top;
  padding: 0.5vw;
  font-weight: bold;
}

.mission-paragraph {
  font-size: 0.6em;
  line-height: 1.5em;
}