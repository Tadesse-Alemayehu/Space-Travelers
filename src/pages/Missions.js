import { useDispatch, useSelector } from 'react-redux';
import { useEffect } from 'react';
import { getMissions } from '../Redux/Missions/Mission';
import '@testing-library/jest-dom';
import Mission from '../components/Mission';
import './Missions.scss';

export default function Missions() {
  const dispatch = useDispatch();
  const Missions = useSelector((state) => state.Missions);
  useEffect(() => {
    if ( Missions.length <= 0)  {
      dispatch(getMissions());
    }
  }, []);

  return (
    <div data-testid="mission-test">
      <table>
        <thead>
          <tr>
            <th>Missions</th>
            <th>Description</th>
            <th>Status</th>
            <th></th>
          </tr>
        </thead>
        <tbody>
          {Missions.map((mission) => (
            <Mission
              key={mission.mission_id}
              id={mission.mission_id}
              reserved={mission.reserved}
              mission={mission}
            />
          ))}
        </tbody>
      </table>
    </div>
  )
}
