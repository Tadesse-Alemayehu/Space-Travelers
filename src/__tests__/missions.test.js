import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import store from '../Redux/configureStore';
import Missions from '../pages/Missions';

it('Check if Missions page renders correctly', () => {
  const tree = renderer.create(
    <Provider store={store}>
      <Missions />
    </Provider>,
  );
  expect(tree).toMatchSnapshot();
});

afterEach(() => {
  jest.clearAllMocks();
});
