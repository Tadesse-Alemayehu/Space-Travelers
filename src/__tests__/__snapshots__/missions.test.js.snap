// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check if Missions page renders correctly 1`] = `
<div>
  <table>
    <thead>
      <tr>
        <th>
          Missions
        </th>
        <th>
          Description
        </th>
        <th>
          Status
        </th>
        <th />
      </tr>
    </thead>
    <tbody />
  </table>
</div>
`;
